dns_sd_windows.c

```diff
@@ -0,0 +1,257 @@
+		if (adapter->TunnelType == TUNNEL_TYPE_TEREDO)
+			continue;
+		if (adapter->OperStatus != IfOperStatusUp)
+			continue;
+
+		for (IP_ADAPTER_UNICAST_ADDRESS* unicast = adapter->FirstUnicastAddress; unicast;
+			unicast = unicast->Next) {
+			if (unicast->Address.lpSockaddr->sa_family == AF_INET) {
+				struct sockaddr_in* saddr = (struct sockaddr_in*)unicast->Address.lpSockaddr;
+				if ((saddr->sin_addr.S_un.S_un_b.s_b1 != 127) ||
+					(saddr->sin_addr.S_un.S_un_b.s_b2 != 0) ||
+					(saddr->sin_addr.S_un.S_un_b.s_b3 != 0) ||
+					(saddr->sin_addr.S_un.S_un_b.s_b4 != 1)) {
+					if (num_sockets < max_sockets) {
+						int sock = mdns_socket_open_ipv4(saddr);
+						if (sock >= 0) {
+							sockets[num_sockets++] = sock;
+						}
+					}
+				}
+			}
+#ifdef HAVE_IPV6
+			else if (unicast->Address.lpSockaddr->sa_family == AF_INET6) {
+				struct sockaddr_in6* saddr = (struct sockaddr_in6*)unicast->Address.lpSockaddr;
+				static const unsigned char localhost[] = { 0, 0, 0, 0, 0, 0, 0, 0,
+														  0, 0, 0, 0, 0, 0, 0, 1 };
+				static const unsigned char localhost_mapped[] = { 0, 0, 0,    0,    0,    0, 0, 0,
+																 0, 0, 0xff, 0xff, 0x7f, 0, 0, 1 };
+				if ((unicast->DadState == NldsPreferred) &&
+					memcmp(saddr->sin6_addr.s6_addr, localhost, 16) &&
+					memcmp(saddr->sin6_addr.s6_addr, localhost_mapped, 16)) {
+					if (num_sockets < max_sockets) {
+						int sock = mdns_socket_open_ipv6(saddr);
+						if (sock >= 0) {
+							sockets[num_sockets++] = sock;
+						}
+					}
+				}
+			}
+#endif
+		}
+	}
+
+	free(adapter_address);
+
+	for (int isock = 0; isock < num_sockets; ++isock) {
+		unsigned long param = 1;
+		ioctlsocket(sockets[isock], FIONBIO, &param);
+	}
+
+	return num_sockets;
+}
+
+
+static int
+query_callback(int sock, const struct sockaddr* from, size_t addrlen,
+	mdns_entry_type_t entry, uint16_t transaction_id,
+	uint16_t rtype, uint16_t rclass, uint32_t ttl,
+	const void* data, size_t size, size_t offset, size_t length,
+	void* user_data) {
+
+
+	char addrbuffer[64];
+	char servicebuffer[64];
+	char namebuffer[256];
+
+	struct dns_sd_discovery_data* dd = (struct dns_sd_discovery_data*)user_data;
+	if (dd == NULL) {
+		IIO_ERROR("DNS SD: Missing info structure. Stop browsing.\n");
+		goto quit;
+	}
+
+	if (rtype != MDNS_RECORDTYPE_SRV)
+		goto quit;
+
+	getnameinfo((const struct sockaddr*)from, (socklen_t)addrlen,
+		addrbuffer, NI_MAXHOST, servicebuffer, NI_MAXSERV,
+		NI_NUMERICSERV | NI_NUMERICHOST);
+
+	mdns_record_srv_t srv = mdns_record_parse_srv(data, size, offset, length,
+		namebuffer, sizeof(namebuffer));
+	IIO_DEBUG("%s : SRV %.*s priority %d weight %d port %d\n",
+		addrbuffer,
+		MDNS_STRING_FORMAT(srv.name), srv.priority, srv.weight, srv.port);
+
+	// Go to the last element in the list
+	while (dd->next != NULL)
+		dd = dd->next;
+
+	if (srv.name.length > 1)
+	{
+		dd->hostname = malloc(srv.name.length);
+		iio_strlcpy(dd->hostname, srv.name.str, srv.name.length);
```

Comment:
why remove the three lines, and  
```
dd->hostname = strndup(srv.name.str, srv.name.length);
```
With strndup, If s is longer than len, only len bytes are copied, and a terminating null byte  ('\0')  is  added.

but we should also check that it is null, and handle the error if it doesn't work. (the malloc fails)

Reply:
MSVC does not have strndup, only strdup which doesn't cut it .. We can use strdup and nullterminate it. I can just leave it like this and add the NULL check.

Reply:
Ok - we have iio_strdup - I would just use that., but we still need the fail / null checking...

Reply:
The string is not null terminated. iio_strdup is bascially a malloc and memcopy which does not guarantee null termination
I used malloc and iio_strlcpy (which guarantees nulltermination). Also added errorcheck on malloc.

I don't think we can get away with one operation here ..

Reply:
Ok - thanks